src/test/editor-service/editor-server-test.ts

```diff
@@ -0,0 +1,155 @@
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {assert} from 'chai';
+import * as child_process from 'child_process';
+import * as path from 'path';
+import * as split from 'split';
+
+import {invertPromise} from '../test-utils';
+
+const pathToServer =
+    path.join(__dirname, '../../editor-service/editor-server.js');
+
+suite('RemoteEditorService', function() {
+  /**
+   * These are the tests. We run these tests using a few different ways of
+   * communicating with the server.
+   */
+  function editorServiceInterfaceTests(
+      sendRequest: (request: any) => Promise<void>,
+      getNextResponse: (expectedId: number) => Promise<any>) {
+    const initMessage = {
+      kind: 'init',
+      basedir: `${path.join(__dirname, 'static')}`
+    };
+    const getWarningsMessage = {
+      kind: 'getWarningsFor',
+      localPath: 'malformed.html'
+    };
+
+
+    test('can create and initialize', async function() {
+      await sendRequest({id: 0, value: initMessage});
+      const response = await getNextResponse(0);
+      assert.deepEqual(response, undefined);
+    });
+    test('initializing twice is an error', async function() {
+      await sendRequest({id: 0, value: initMessage});
+      await getNextResponse(0);
+      await sendRequest({id: 1, value: initMessage});
+      const errorMessage = await invertPromise(getNextResponse(1));
+      assert.equal(errorMessage, 'Already initialized!');
+    });
+    test('the first request must be initialization', async function() {
+      await sendRequest({id: 0, value: getWarningsMessage});
+      const errorMessage = await invertPromise(getNextResponse(0));
+      assert.equal(
+          errorMessage,
+          `Must send an 'init' message before any others. Received ` +
+              `'getWarningsFor' message before 'init'.`);
+    });
+    const testName = 'can perform editor service functions once initialized';
+    test(testName, async function() {
+      await sendRequest({id: 0, value: initMessage});
+      await getNextResponse(0);
+      await sendRequest({id: 1, value: getWarningsMessage});
+      const warnings = await getNextResponse(1);
+      assert.deepEqual(warnings, [{
+                         code: 'parse-error',
+                         message: 'Unexpected token <',
+                         severity: 0,
+                         sourceRange: {
+                           file: 'malformed.html',
+                           start: {line: 266, column: 0},
+                           end: {line: 266, column: 0}
+                         }
+                       }]);
+    });
+  }
+
+  let suiteName =
+      'from node with child_process.fork() and process.send() for IPC';
```

Comment:
+1. This was mostly to work around an issue with clang-format, which was super indenting test bodies very deeply when test names wrapped.